From a6fa8816f9b50dd6995ef666fce99a96eb0f75a7 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Mon, 15 Jun 2020 14:46:31 +0300
Subject: [PATCH] [SW-14252] zephyr: create little endian binaries

This patch creates little endian binaries for zephyr iccm and dccm
binaries which will be used for FIT creation.

Change-Id: Ifccee94e8f7a8806f9127794d01bb344afedb9d4
Signed-off-by: Sanyog Kale <skale@habana.ai>
---

diff --git a/board/habanalabs/common/post-image-spi-arc_global.sh b/board/habanalabs/common/post-image-spi-arc_global.sh
index f823eba..4780f6b 100755
--- a/board/habanalabs/common/post-image-spi-arc_global.sh
+++ b/board/habanalabs/common/post-image-spi-arc_global.sh
@@ -15,6 +15,10 @@
 
 ZEPHYR_PID_DIR=${BR2_EXTERNAL}/output/${DEV_NAME}_arc_gnu/build/zephyr-custom/build-pid
 ZEPHYR_MGMT_DIR=${BR2_EXTERNAL}/output/${DEV_NAME}_arc_gnu/build/zephyr-custom/build-mgmt
+ZEPHYR_PID_ICCM_BINFILE=${ZEPHYR_PID_ICCM_BINFILE:-zephyr-pid-iccm.bin}
+ZEPHYR_PID_DCCM_BINFILE=${ZEPHYR_PID_DCCM_BINFILE:-zephyr-pid-dccm.bin}
+ZEPHYR_MGMT_ICCM_BINFILE=${ZEPHYR_MGMT_ICCM_BINFILE:-zephyr-mgmt-iccm.bin}
+ZEPHYR_MGMT_DCCM_BINFILE=${ZEPHYR_MGMT_DCCM_BINFILE:-zephyr-mgmt-dccm.bin}
 
 PRIME_IMAGE=0
 SECOND_IMAGE=1
@@ -326,6 +330,16 @@
 	cp ${FILEPATH}/zephyr/zephyr-dccm.bin ${OUTPUT_DIR}/zephyr-${PREFIX}-dccm.bin
 }
 
+function swap_file {
+	local INF=$1
+	local OUTF=$2
+
+	# SWAP needs this to be 4B aligned in size
+	dd status=none conv=sync bs=4 if=${INF} > ${OUTF}.tmp
+	# Sadly --endian=big is not swapping for -tx1 so I used some gawk magic
+	od -Anone -v -tx1 -w4 ${OUTF}.tmp | gawk '{ print $4 $3 $2 $1 }' | xxd -r -p > ${OUTF}
+}
+
 pushd ${OUTPUT_DIR}
 
 	hl_copy_binaries ${OUTPUT_DIR}
@@ -423,4 +437,24 @@
 
 	# for otp flash
 	hl_pldm_flash_hex spi-otp-flash ${DEV_NAME}-spi-otp-flash.bin
+
+       dd status=none conv=sync bs=8 if=${ZEPHYR_PID_ICCM_BINFILE} > ${DEV_NAME}-zephyr-pid-iccm.bin.le
+       # Swap zephyr pid iccm
+       swap_file ${DEV_NAME}-zephyr-pid-iccm.bin.le ${DEV_NAME}-zephyr-pid-iccm.bin.be
+       ZEPHYR_PID_ICCM_BINFILE=${DEV_NAME}-zephyr-pid-iccm.bin.be
+
+       dd status=none conv=sync bs=8 if=${ZEPHYR_PID_DCCM_BINFILE} > ${DEV_NAME}-zephyr-pid-dccm.bin.le
+       # Swap zephyr pid iccm
+       swap_file ${DEV_NAME}-zephyr-pid-dccm.bin.le ${DEV_NAME}-zephyr-pid-dccm.bin.be
+       ZEPHYR_PID_DCCM_BINFILE=${DEV_NAME}-zephyr-pid-dccm.bin.be
+
+       dd status=none conv=sync bs=8 if=${ZEPHYR_MGMT_ICCM_BINFILE} > ${DEV_NAME}-zephyr-mgmt-iccm.bin.le
+       # Swap zephyr pid iccm
+       swap_file ${DEV_NAME}-zephyr-mgmt-iccm.bin.le ${DEV_NAME}-zephyr-mgmt-iccm.bin.be
+       ZEPHYR_MGMT_ICCM_BINFILE=${DEV_NAME}-zephyr-pid-iccm.bin.be
+
+       dd status=none conv=sync bs=8 if=${ZEPHYR_MGMT_DCCM_BINFILE} > ${DEV_NAME}-zephyr-mgmt-dccm.bin.le
+       # Swap zephyr pid iccm
+       swap_file ${DEV_NAME}-zephyr-mgmt-dccm.bin.le ${DEV_NAME}-zephyr-mgmt-dccm.bin.be
+       ZEPHYR_MGMT_DCCM_BINFILE=${DEV_NAME}-zephyr-mgmt-dccm.bin.be
 popd
